beater/interceptors/metrics.go

```diff
@@ -0,0 +1,67 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package interceptors
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+
+	"github.com/elastic/apm-server/beater/request"
+	"github.com/elastic/beats/v7/libbeat/monitoring"
+)
+
+// MetricsMonitoringKeys are the keys used when incrementing interceptor counters.
+var MetricsMonitoringKeys = []request.ResultID{
+	request.IDRequestCount, request.IDResponseCount, request.IDResponseErrorsCount,
+	request.IDResponseValidCount, request.IDEventReceivedCount, request.IDEventDroppedCount,
+}
+
+// Metrics increments the counters in map m according to the passed in error,
+// the result of a gRPC method call. The monitoring registry map must be passed
+// into the request context during execution. The returned function implements
+// grpc.UnaryServerInterceptor.
+func Metrics(
+	registries map[string]map[request.ResultID]*monitoring.Int,
+) grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		m, prs := registries[info.FullMethod]
+		if !prs {
+			return handler(ctx, req)
```

Comment:
If a new gRPC method is introduced and the incoming `registries` argument isn't updated, it would probably be good to indicate this somehow. Should it be logged? Or what?